designs/package_during_deploy.md

```diff
@@ -0,0 +1,175 @@
+
+Addition of extra parameters that are currently supported by `sam package` over to `sam deploy` with all of them being optional.
+
+Additional parameters that need to be on `sam deploy` that are not on `sam package`.
+
+* `--metadata`
+* `--use-json`
+
+Parameters that dont need to be added.
+
+* `--output-template-file`
+ * 	An explicit `output-template-file` is created on the fly during packaging in the deploy phase.
+
+If the expectation is to package and deploy in one command, One can now do.
+
+`sam deploy --stack-name sam-package-on-deploy --capabilities CAPABILITY_IAM --s3-bucket sam-package-bucket`
+
+There is no explicit need to pass in a `--template-file` parameter, if one is not passed in it to defaults to trying to find the template.yaml that was generated by `sam build` which is located at `.aws-sam/build/template.yaml`
+
+The old method of deploying pre-packaged artifacts will continue to work as before.
+
+* `sam deploy --template-file packaged.yaml --stack-name sam-package-on-deploy --capabilities  CAPABILITY_IAM`
+
+Future of `sam package`?
+---------------------
+
+* `sam package` will continue to exist in the state it is today and will continue to be improved upon separately.
+
+Success criteria for the change
+-------------------------------
+
+User do not require to run `sam package` as part of their author test loop, except for CI/CD purposes, where `sam package` can be run and the packaged template file can be passed to cloudformation deploy actions.
+
+
+Out-of-Scope
+------------
+
+The s3 bucket where the packaged artifacts go is not abstracted in this design. In the future, the s3 bucket could be specified via a configuration file.
+
+This is currently in design in : https://github.com/awslabs/aws-sam-cli/pull/1503
+
+User Experience Walkthrough
+---------------------------
+
+`sam build` -> `sam deploy`
+
+`sam build` -> `sam package` -> `sam deploy`
+
+Provide identical experiences in terms of a deploying the same stack, with exactly same artifacts.
+
+
+Implementation
+==============
+
+CLI Changes
+-----------
+
+* Add new arguments `--metadata`, `--use-json` and modify existing `--template-file` to look for a default `template.yaml`  that exists under `.aws-sam/build/`
+
+### Breaking Change
+
+* Not a breaking change , but there are optional behavorial changes that a user can subscribe into by supplying a non-packaged template file and an s3 bucket.
+
+Design
+------
+
+* Changes to Deploy command's click options
+* Attempt to package on every deploy if an appropriate s3 bucket is specified and deploy using the output template file during package. 
+* If a pre-packaged template is specified, an attempt to package does not change the template and the same template is used for deploy.
+* The parameters that share the same name across package and deploy are collapsed together. eg: `--kms-key-id` , if a kms-key-id is specified that same key is used across both packaging and deploy purposes.
+
+`.samrc` Changes
```

Comment:
We should update the template in the sam config so this gets reflected in new designs going forward. Just noticing this now but not really apart of this pr directly.